### PR TITLE
DRILL-3714: Avoid cascading disconnection when a single connection is broken

### PR DESCRIPTION
**Pull request by jacques-n**
DRILL-3714: Avoid cascading disconnection when a single connection is broken

- Move the coordination id management to be connection level instead of RpcBus level
- Rename CoordinationQueue to a more appropriate name: RequestIdMap
- Simplify locking and memory overhead of RequestIdMap. It used to be that this would accessed by a large number of threads concurrently. We modified the behavior so that it is only accessed by two threads at most. Rather than have memory overhead of ConcurrentHashMap, switch to simple locking approach since contention should be low.
- Update all methods associated with coordination to improve names as well as add javadocs. Move these methods to the RemoteConnection.
- Consolidate the two different close handlers into a single, ordered close handler managed inside the connection.
- Add better javadoc around the close method of RemoteConnection
- Add some preconditions checks.
- Update the HPPC version in the base memory module since it conflicts with the one in the java-exec module.

### CODE REVIEW

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RemoteConnection.java [hunk 1]
```diff
@@ -121,17 +115,72 @@ public void disable() {
...
     public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
-//      logger.debug("Channel writability changed.", ctx.channel().isWritable());
       writeManager.setWritable(ctx.channel().isWritable());
       ctx.fireChannelWritabilityChanged();
     }
 
   }
 
+  /**
+   * For incoming messages, remove the outcome listener and return it. Can only be done once per coordinationId.
```
Comment (adeneche) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RemoteConnection.java]: > can only be done once per coordinationId

Could this be slightly misleading ? `requestIdMap` recycles coordinationIds, so technically this can be called **multiple times** per coordinationId if the Drillbit runs long enough.

**Comment by adeneche** [comment]
apart from a small javadoc change, LGTM

+1

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/BasicClient.java [hunk 2]
```diff
@@ -282,15 +283,19 @@ public void interrupted(final InterruptedException ex) {
 
   private class ClientHandshakeHandler extends AbstractHandshakeHandler<HANDSHAKE_RESPONSE> {
 
-    public ClientHandshakeHandler() {
+    private final R connection;
+
+    public ClientHandshakeHandler(R connection) {
       super(BasicClient.this.handshakeType, BasicClient.this.handshakeParser);
+      Preconditions.checkNotNull(connection);
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/BasicClient.java]: `this.connection = Preconditions.checkNotNull(connection);`

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java [hunk 3]
```diff
@@ -20,51 +20,82 @@
...
+ * Manages the creation of rpc futures for a particular socket <--> socket
+ * connection. Generally speaking, there will be two threads working with this
+ * class (the socket thread and the Request generating thread). Synchronization
+ * is simple with the map being the only thing that is protected. Everything
+ * else works via Atomic variables.
  */
-public class CoordinationQueue {
-  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(CoordinationQueue.class);
+class RequestIdMap {
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(RequestIdMap.class);
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: private

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java [hunk 4]
```diff
@@ -20,51 +20,82 @@
...
+
+  private final AtomicInteger value = new AtomicInteger();
+  private final AtomicBoolean acceptMessage = new AtomicBoolean(true);
 
-  private final PositiveAtomicInteger circularInt = new PositiveAtomicInteger();
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: Remove PositiveAtomicIneteger class.

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java [hunk 5]
```diff
@@ -20,51 +20,82 @@
...
+  private class Closer implements IntObjectProcedure<RpcOutcome<?>> {
+    final RpcException exception;
+
+    public Closer(RpcException exception) {
+      this.exception = exception;
+    }
+
+    @Override
+    public void apply(int key, RpcOutcome<?> value) {
+      try{
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: Inconsistent spacing here and below.

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java [hunk 6]
```diff
@@ -20,51 +20,82 @@
...
-    if (old != null) {
-      throw new IllegalStateException(
-          "You attempted to reuse a coordination id when the previous coordination id has not been removed.  This is likely rpc future callback memory leak.");
+    final Object old;
+    synchronized (map) {
+      Preconditions.checkArgument(acceptMessage.get(),
+          "Attempted to send a message when connection is no longer valid.");
+      old = map.put(i, future);
     }
+    Preconditions.checkArgument(old == null,
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: Not required, since numbers are no longer reused?

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java [hunk 7]
```diff
@@ -158,22 +157,16 @@ public ChannelClosedHandler(C clientConnection, Channel channel) {
...
-      String msg;
+      final String msg;
+
       if(local!=null) {
         msg = String.format("Channel closed %s <--> %s.", local, remote);
       }else{
         msg = String.format("Channel closed %s <--> %s.", future.channel().localAddress(), future.channel().remoteAddress());
       }
 
-      if (RpcBus.this.isClient()) {
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java]: `isClient` method is no longer used. Remove the method.

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java [hunk 8]
```diff
@@ -261,6 +251,7 @@ public void execute(Runnable command) {
 
     public InboundHandler(C connection) {
       super();
+      Preconditions.checkNotNull(connection);
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RpcBus.java]: `this.connection = Preconditions.checkNotNull(connection);`

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java [hunk 9]
```diff
@@ -20,51 +20,82 @@
...
+ * class (the socket thread and the Request generating thread). Synchronization
+ * is simple with the map being the only thing that is protected. Everything
+ * else works via Atomic variables.
  */
-public class CoordinationQueue {
-  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(CoordinationQueue.class);
+class RequestIdMap {
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(RequestIdMap.class);
+
+  private final AtomicInteger value = new AtomicInteger();
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: How about `coordinationIdCounter` and `isOpen`?

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java [hunk 10]
```diff
@@ -20,51 +20,82 @@
...
+      synchronized (map) {
+        map.forEach(new Closer(e));
+        map.clear();
       }
-      for (RpcOutcome<?> f : map.values()) {
-        f.setException(e);
+    }
+  }
+
+  private class Closer implements IntObjectProcedure<RpcOutcome<?>> {
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: Better class name? `SetExceptionProcedure`?

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java [hunk 11]
```diff
@@ -20,51 +20,82 @@
...
+      RemoteConnection connection) {
+    int i = value.incrementAndGet();
     RpcListener<V> future = new RpcListener<V>(handler, clazz, i, connection);
-    Object old = map.put(i, future);
-    if (old != null) {
-      throw new IllegalStateException(
-          "You attempted to reuse a coordination id when the previous coordination id has not been removed.  This is likely rpc future callback memory leak.");
+    final Object old;
+    synchronized (map) {
+      Preconditions.checkArgument(acceptMessage.get(),
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: Make this check first statement in the method?

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java [hunk 12]
```diff
@@ -84,7 +115,7 @@ public void operationComplete(ChannelFuture future) throws Exception {
       if (!future.isSuccess()) {
         removeFromMap(coordinationId);
         if (future.channel().isActive()) {
-           throw new RpcException("Future failed") ;
+          throw new RpcException("Future failed");
```
Comment (sudheeshkatkam) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: Since the future did not succeed, should this `setException(future.cause())`? There would be no outcome for the `handler` otherwise, right?

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java (hunk 4 above)
Comment (adeneche) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: I didn't notice this. I have a question about this change:

so we still recycle coordinationIds, it's just that we don't care if they become negative if the Drillbit runs long enough. Is this correct ?

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RemoteConnection.java (hunk 1 above)
Comment (adeneche) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RemoteConnection.java]: > can only be done once per coordinationId

Could this be slightly misleading ? `requestIdMap` recycles coordinationIds, so technically this can be called **multiple times** per coordinationId if the Drillbit runs long enough.

**Comment by sudheeshkatkam** [comment]
Overall change looks good; I have some minor comments.

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java (hunk 4 above)
Comment (jacques-n) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: Correct. Originally (long ago), this was structured as an array. However, we're using a map now so it doesn't matter.

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java (hunk 6 above)
Comment (jacques-n) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: This is an assertion to ensure that there isn't a bug some place.

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java (hunk 11 above)
Comment (jacques-n) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: I prefer to spend as little time in the synchronized block as possible. If we move this up we still need to check in the synchronized block (e.g. we could make this double-checked).

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java (hunk 12 above)
Comment (jacques-n) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/RequestIdMap.java]: This has existed since this commit (Sept 2013): https://github.com/apache/drill/commit/3f101aabd05f174e8bfc9dbdd2590066a1e937f4

I'm not inclined to change it in this patch as it seems out of scope. The change probably makes sense but let's explore in a different jira to ensure we're thinking through the ramifications.

File: exec/rpc/src/main/java/org/apache/drill/exec/rpc/BasicClient.java (hunk 2 above)
Comment (jacques-n) [inline exec/rpc/src/main/java/org/apache/drill/exec/rpc/BasicClient.java]: Note that I dislike this construct so I didn't include this change.
